analysis/quantifier-comparisons/04-cell-level-benchmarking-metrics.Rmd

```diff
@@ -0,0 +1,719 @@
+  purrr::map_df(celldata_to_df, .id = "quant_id")
+
+# combine all the data frames into one
+cell_qc <- dplyr::bind_rows(
+  alevin = alevin_cell_qc,
+  alevin_fry = alevin_fry_cell_qc,
+  alevin_fry_sketch = alevin_fry_sketch_cell_qc,
+  alevin_fry_unfiltered = alevin_fry_unfiltered_cell_qc,
+  alevin_fry_unfiltered_sketch = alevin_fry_unfiltered_sketch_cell_qc,
+  kallisto = kallisto_cell_qc,
+  cellranger = cellranger_cell_qc,
+  .id = "tool"
+) %>%
+  dplyr::left_join(quant_info,
+                   by = c("tool" = "tool", 
+                          "quant_id" = "quant_dir")) %>%
+  # remove extraneous run from 119 with txome index since we have the run with the spliced index
+  dplyr::filter(quant_id != "SCPCR000119-txome_k31") %>%
+  # remove duplicate runs of alevin-fry --knee on 118, 119
+  dplyr::filter(quant_id !="SCPCR000118-spliced_intron_txome_k31-salign") %>%
+  dplyr::filter(quant_id != "SCPCR000119-spliced_intron_txome_k31-salign")
+                  
+```
+
+
+## Comparison of QC Metrics across all tools
+
+### Per Cell QC
+
+Let's first start by taking a look at some general information across all of our tools and how consistent each of these metrics is for all four of the samples. Here are the metrics we want to look at: 
+
+  * Total number of cells detected
+  * Mitochondrial content/ cell
+  * Number of UMIs/cell
+  * Number of Genes/cell 
+
+
+```{r}
+## make new data frame with mean of cells detected and standard dev
+cells_detected_stats <- cell_qc %>%
+  group_by(sample, tool) %>%
+  summarise(cells_detected = mean(cells_detected))
+cells_detected_stats
+
+## first look at cells detected
+ggplot(cells_detected_stats, aes(x = tool, y = cells_detected)) +
+  geom_jitter(mapping = aes(color = sample)) +
+  geom_violin() +
+  theme_classic() +
+  ylab("Total Cells Detected") + 
+  xlab("") + 
+  theme(axis.text.x = element_text(angle = 45, hjust = 1))
+```
+Based on this, it looks like alevin-fry using the `--knee-distance` and kallisto do not handle single-nuclei samples well. 
+In contrast, the alevin-fry `--unfiltered-pl` samples were filtered using `DropletUtils::emptyDrops` and they have much lower numbers for the same single-nuclei samples. 
+
+Let's take a look at the mitochondrial content per cell across the tools and see if we see variation there. 
+
+
+```{r fig.height = 10, fig.width = 10}
+ggplot(cell_qc, aes(x = tool, y = subsets_mito_percent, fill = tool)) + 
+  geom_boxplot() + 
+  facet_grid(sample ~ index_type) + 
+  theme_classic() + 
+  ylab("% Mito /Cell") + 
+  theme(axis.ticks.x = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1))
+```
+It looks like SCPCR000003 is a very poor sample across all tools and has a lot of dead cells. 
+It might be a good idea to remove that from our comparisons and not consider this for benchmarking. 
+Additionally, it looks like we can see that SCPCR000118 (snRNAseq) also has some variation, but the other samples tend to be fairly consistent. 
```

Comment:
This is super odd, and seems like a potential problem for alevin-fry with the the default settings. mito genes should be essentially nonexistent in nuclei, so I am surprised that we are seeing so much variability there! A question I have is whether those are the same cells, or if alevin/-fry are not throwing out some droplets that should have been discarded...

Reply:
Really interesting analyses @allyhawkins!  I'm Tagging @k3yavi and @mohsenzakeri here to take a look at these samples from our end.

Reply:
Hey guys, thanks for maintaining such a great resource. Just to add to Rob's point here, we also had similar observation about the mito/ribo genes, for single-cell data (not single-nuclei) having high counts compared to other tools. It has been a while but if I remember correctly these group of genes share highly repetitive genomic sequences which almost all the tools ignore in their quantification methods (some tools have EM related flags but they are disabled by default), which is enabled by default in alevin. I'd recommend checking the uniqueness ratio of these outlier genes to get a better sense of why alevin is showing more counts for them compared to other tools in non-ambiguous modes.

Reply:
Thanks for the insight on this topic. What's interesting is that when we look at the same sample run on alevin-fry with --sketch enabled we see a decrease in the percentage of reads mapping to mito/ribo genes. 
It seems like it is a consequence of using alevin/alevin-fry in its default mode with selective alignment rather than pseudoalignment? Does using the --sketch flag in alevin-fry change the default settings in Alevin to not use the EM flags? What are the benefits to having that enabled vs. disabled?

Reply:
Hi Ally.  So, this is certainly an interesting observation.  It's _not_ the case that `--sketch` mode turns off the EM automatically or some such.  The resolution mode is independent of how the records make their way into the RAD file.  Here, my best guess is that selective-alignment is just more sensitive, and is somehow finding alignments to these genes that either (a) are good enough given the default filtering criterion for alignments (and no other equally good alignment is being found to other genes) or (b) are as good as the alignments that it is finding with respect to some other gene --- leading to this alignment being multi-mapping, and the corresponding counts being split during the `full` resolution.  The selective-alignment algorithm verifies all of the alignments it finds with "normal" dynamic programming scoring.  

So this means that it must be finding alignments for fragments to these ribo genes that are good enough to pass the filtering criteria (which are, themselves, modifiable).  As @k3yavi points out, it's not immediately obvious if reads are mapping to these genes in other methods or not, as they will simply discard gene multimapping reads rather than try to resolve them.  In the case of Cell Ranger, one could look at one of the BAM files and see if alignments exist to these genes, but are simply not being counted.  Likewise, one could run alevin-fry in `--cr-like` mode and see how that changes the abundance of those genes in these samples.  Finally, I'll mention that the two mapping algorithms (with and without sketch) are certainly different.  While selective alignment will validate the mapping with a score, and return all (and only) those places where the mapping yields an equally best score, the role of the "score" in `--sketch` mode is instead played by what set of k-mers are searched and found.  As such, it's always possible that there may be different alignments with equally valid alignment scores, but only one (or a subset) the mappings is found in sketch mode because of the heuristics used to extract the k-mer set that is eventually tested.